File: src/execution/paper_executor.go
package execution

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// Position 持仓跟踪结构
type Position struct {
	TokenID       string
	Side          string
	Quantity      float64
	AvgEntryPrice float64
	TotalInvested float64
}

// PaperExecutor 模拟交易执行器
// 不使用真实资金进行交易模拟
type PaperExecutor struct {
	balance        float64
	initialBalance float64
	positions      map[string]*Position
	tradeHistory   []interface{}
	pendingOrders  map[string]TradeParams
	filledOrders   map[string]OrderStatus
	tokenSideMap   map[string]string
	mu             sync.RWMutex
}

// NewPaperExecutor 创建新的模拟执行器
func NewPaperExecutor(initialBalance float64) *PaperExecutor {
	fmt.Printf("\n📝 PAPER TRADING PORTFOLIO INITIALIZED\n")
	fmt.Printf("   💰 Initial Cash: $%.2f\n", initialBalance)
	fmt.Printf("   ------------------------------------------\n\n")

	return &PaperExecutor{
		balance:        initialBalance,
		initialBalance: initialBalance,
		positions:      make(map[string]*Position),
		tradeHistory:   make([]interface{}, 0),
		pendingOrders:  make(map[string]TradeParams),
		filledOrders:   make(map[string]OrderStatus),
		tokenSideMap:   make(map[string]string),
	}
}

// GetBalance 获取当前余额
func (p *PaperExecutor) GetBalance() (float64, error) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.balance, nil
}

// PlaceOrder 下订单
func (p *PaperExecutor) PlaceOrder(params TradeParams) (string, error) {
	p.mu.Lock()
	defer p.mu.Unlock()

	// 生成订单 ID
	orderID := fmt.Sprintf("PAPER-%d-%d", time.Now().Unix()%1000000, rand.Intn(100))

	// 存储为待处理订单
	p.pendingOrders[orderID] = params

	orderType := ""
	if params.Type != "" {
		orderType = fmt.Sprintf("[%s]", params.Type)
	}

	fmt.Printf("\n📝 [PAPER] Order Placed (Pending): %s %.2f @ $%.2f %s (ID: %s)\n",
		params.Side, params.Size, params.Price, orderType, orderID)

	return orderID, nil
}

// CancelAll 取消所有订单
func (p *PaperExecutor) CancelAll() error {
	p.mu.Lock()
	defer p.mu.Unlock()

	p.pendingOrders = make(map[string]TradeParams)
	fmt.Println("   [PAPER] All pending orders cleared.")
	return nil
}

// CancelOrder 取消指定订单
func (p *PaperExecutor) CancelOrder(orderID string) error {
	p.mu.Lock()
	defer p.mu.Unlock()

	if _, exists := p.pendingOrders[orderID]; exists {
		delete(p.pendingOrders, orderID)
		fmt.Printf("   [PAPER] Cancelled Pending Order %s\n", orderID)
	} else {
		fmt.Printf("   [PAPER] Cancelled Order %s (Already filled or non-existent)\n", orderID)
	}
	return nil
}

// GetOrderStatus 获取订单状态
func (p *PaperExecutor) GetOrderStatus(orderID string) (OrderStatus, error) {
	p.mu.RLock()
	defer p.mu.RUnlock()

	// 检查是否已成交
	if filled, exists := p.filledOrders[orderID]; exists {
		return filled, nil
	}

	// 检查是否仍在待处理
	if _, exists := p.pendingOrders[orderID]; exists {
		return OrderStatus{Matched: 0, Cancelled: false, AvgFillPrice: 0}, nil
	}

	// 未找到 - 假设已取消
	return OrderStatus{Matched: 0, Cancelled: true, AvgFillPrice: 0}, nil
}

// GetPositions 获取持仓
func (p *PaperExecutor) GetPositions(tokenIDYes, tokenIDNo string) (Positions, error) {
	p.mu.RLock()
	defer p.mu.RUnlock()

	posYes := p.positions[tokenIDYes]
	posNo := p.positions[tokenIDNo]

	yesQty := 0.0
	noQty := 0.0

	if posYes != nil {
		yesQty = posYes.Quantity
	}
	if posNo != nil {
		noQty = posNo.Quantity
	}

	return Positions{Yes: yesQty, No: noQty}, nil
}

// GetOpenOrders 获取开放订单
func (p *PaperExecutor) GetOpenOrders(marketSlug string) ([]OpenOrder, error) {
	p.mu.RLock()
	defer p.mu.RUnlock()

	orders := make([]OpenOrder, 0)

	for orderID, params := range p.pendingOrders {
		side := p.tokenSideMap[params.TokenID]
		if side == "" {
			side = "YES" // 默认为 YES
		}

		orders = append(orders, OpenOrder{
			ID:    orderID,
			Side:  side,
			Price: fmt.Sprintf("%.2f", params.Price),
			Size:  params.Size,
		})
	}

	return orders, nil
}

// GetAddress 获取用户地址 (模拟交易返回空地址)
func (p *PaperExecutor) GetAddress() (string, error) {
	return "0x0000000000000000000000000000000000000000", nil
}

// RedeemPositions 赎回持仓 (模拟交易返回虚拟哈希)
func (p *PaperExecutor) RedeemPositions(conditionID string) (string, error) {
	fmt.Printf("📝 [PAPER] Redemption simulated for condition %s...\n", conditionID[:10])
	return fmt.Sprintf("PAPER-REDEEM-%d", time.Now().Unix()), nil
}

// ExecutePaperFill 手动执行模拟成交
func (p *PaperExecutor) ExecutePaperFill(orderID string, price, quantity float64) error {
	p.mu.Lock()
	defer p.mu.Unlock()

	params, exists := p.pendingOrders[orderID]
	if !exists {
		fmt.Printf("❌ [PAPER] Cannot fill unknown/cancelled order %s\n", orderID)
		return fmt.Errorf("order not found")
	}

	// 模拟延迟
	time.Sleep(50 * time.Millisecond)

	cost := price * quantity

	// 验证资金
	if params.Side == "BUY" && p.balance < cost {
		fmt.Println("❌ [PAPER] REJECTED: Insufficient Funds")
		delete(p.pendingOrders, orderID)
		return fmt.Errorf("insufficient funds")
	}

	// 执行交易
	if params.Side == "BUY" {
		p.balance -= cost
	} else {
		p.balance += cost // 卖出逻辑
	}

	// 更新持仓
	p.updatePosition(TradeParams{
		TokenID: params.TokenID,
		Side:    params.Side,
		Price:   price,
		Size:    quantity,
	})

	// 标记为已成交
	p.filledOrders[orderID] = OrderStatus{
		Matched:      quantity,
		AvgFillPrice: price,
		Cancelled:    false,
	}

	// 从待处理订单中移除
	delete(p.pendingOrders, orderID)

	// 打印交易日志
	p.printTradeLog(orderID, TradeParams{
		TokenID: params.TokenID,
		Side:    params.Side,
		Price:   price,
		Size:    quantity,
	}, cost)

	p.printPortfolioSummary()

	return nil
}

// updatePosition 更新持仓
func (p *PaperExecutor) updatePosition(params TradeParams) {
	existing := p.positions[params.TokenID]

	if existing != nil {
		// 计算加权平均价格
		totalShares := existing.Quantity + params.Size
		totalCost := existing.TotalInvested + params.Price*params.Size

		existing.Quantity = totalShares
		existing.TotalInvested = totalCost
		existing.AvgEntryPrice = totalCost / totalShares
	} else {
		// 新建持仓
		p.positions[params.TokenID] = &Position{
			TokenID:       params.TokenID,
			Side:          params.Side,
			Quantity:      params.Size,
			AvgEntryPrice: params.Price,
			TotalInvested: params.Price * params.Size,
		}
	}
}

// printTradeLog 打印交易日志
func (p *PaperExecutor) printTradeLog(id string, params TradeParams, cost float64) {
	fmt.Printf("\n✅ [PAPER TRADE EXECUTED] #%s\n", id)
	fmt.Printf("   🔹 Side:      %s\n", params.Side)
	fmt.Printf("   🔹 Asset:     ...%s\n", params.TokenID[len(params.TokenID)-10:])
	fmt.Printf("   🔹 Price:     $%.2f\n", params.Price)
	fmt.Printf("   🔹 Size:      %.2f Shares\n", params.Size)
	fmt.Printf("   🔹 Cost:      $%.2f\n", cost)
}

// printPortfolioSummary 打印投资组合摘要
func (p *PaperExecutor) printPortfolioSummary() {
	totalInvested := 0.0
	for _, pos := range p.positions {
		totalInvested += pos.TotalInvested
	}

	totalEquity := p.balance + totalInvested
	profit := totalEquity - p.initialBalance
	profitColor := "🟢"
	if profit < 0 {
		profitColor = "🔴"
	}

	fmt.Printf("\n📊 [PORTFOLIO SUMMARY]\n")
	fmt.Printf("   💵 Cash Balance:    $%.2f\n", p.balance)
	fmt.Printf("   💼 Invested Amount: $%.2f\n", totalInvested)
	fmt.Printf("   📈 Total Equity:    $%.2f\n", totalEquity)
	fmt.Printf("   %s PnL (Unrealized): $%.2f\n", profitColor, profit)

	fmt.Println("   -----------------------------------------------------------")
	fmt.Println("   POSITIONS:")
	if len(p.positions) == 0 {
		fmt.Println("   (Empty)")
	} else {
		for id, pos := range p.positions {
			fmt.Printf("   • ID ...%s | Qty: %.2f | Avg Entry: $%.3f | Cost: $%.2f\n",
				id[len(id)-8:], pos.Quantity, pos.AvgEntryPrice, pos.TotalInvested)
		}
	}
	fmt.Println("   -----------------------------------------------------------\n")
}

File: _upstream/src/execution/PaperExecutor.ts
/**
 * PAPER EXECUTOR
 * 
 * Simulates trading without using real funds.
 * 
 * Features:
 * - Fake $1,000 starting balance
 * - Simulated order placement
 * - Manual order fills (via executePaperFill)
 * - Position tracking
 * - Trade history logging
 * 
 * Use this for:
 * - Testing strategies
 * - Learning how the bot works
 * - Backtesting
 * - Development
 * 
 * ⚠️ Note: Orders don't automatically fill - you must call executePaperFill()
 * to simulate fills. In real trading, fills come from WebSocket.
 * 
 * @class PaperExecutor
 * @implements {IExecutor}
 */
import { IExecutor, TradeParams } from "./IExecutor";

/**
 * Position tracking structure
 * 
 * Tracks:
 * - Token ID
 * - Side (BUY/SELL)
 * - Quantity (shares)
 * - Average entry price
 * - Total invested (cost basis)
 */
interface Position {
  tokenId: string;
  side: "BUY" | "SELL";
  quantity: number;
  avgEntryPrice: number;
  totalInvested: number;
}

export class PaperExecutor implements IExecutor {
  // ============================================================================
  // PROPERTIES
  // ============================================================================
  
  /** Current fake USDC balance */
  private balance: number;
  
  /** Starting balance (for performance tracking) */
  private initialBalance: number;
  
  /** Current positions (tokenId -> Position) */
  private positions: Map<string, Position> = new Map();
  
  /** Trade history (for logging and analysis) */
  private tradeHistory: any[] = [];
  
  /** Pending orders (orders that haven't been filled yet) */
  private pendingOrders: Map<string, TradeParams> = new Map();
  
  /** Filled orders (orders that have been executed) */
  private filledOrders: Map<string, { matched: number; avgFillPrice: number }> =
    new Map();
  
  /**
   * Token ID to side mapping
   * Used for order recovery and position tracking
   */
  private tokenSideMap: Map<string, "YES" | "NO"> = new Map();

  // ============================================================================
  // CONSTRUCTOR
  // ============================================================================
  
  /**
   * Initialize PaperExecutor
   * 
   * Creates a simulated trading account with fake funds.
   * 
   * @param {number} initialBalance - Starting balance (default: $1,000)
   */
  constructor(initialBalance: number = 1000) {
    this.balance = initialBalance;
    this.initialBalance = initialBalance;
    console.log(`\n📝 PAPER TRADING PORTFOLIO INITIALIZED`);
    console.log(`   💰 Initial Cash: $${this.balance.toFixed(2)}`);
    console.log(`   ------------------------------------------\n`);
  }

  async getBalance(): Promise<number> {
    return this.balance;
  }

  async placeOrder(params: TradeParams): Promise<string> {
    // 1. GENERATE ID
    const orderId = `PAPER-${Date.now().toString().slice(-6)}-${Math.floor(
      Math.random() * 100
    )}`;

    // 2. STORE AS PENDING
    // In "Trap" strategy, we place orders that sit in the book.
    // They are NOT filled immediately.
    // Note: FOK/IOC types don't apply to paper trading, but we accept them for interface consistency
    this.pendingOrders.set(orderId, params);

    // 🟢 Track tokenId for order recovery (simplified: assume BUY orders are YES/NO based on context)
    // This is a placeholder - real implementation would need market context
    // For now, we'll infer from the order pattern (all orders are BUY in this strategy)

    const orderType = params.type ? `[${params.type}]` : "";
    console.log(
      `\n📝 [PAPER] Order Placed (Pending): ${params.side} ${
        params.size
      } @ $${params.price.toFixed(2)} ${orderType} (ID: ${orderId})`
    );
    return orderId;
  }

  async cancelAll(): Promise<void> {
    this.pendingOrders.clear();
    console.log("   [PAPER] All pending orders cleared.");
  }

  async cancelOrder(orderId: string): Promise<void> {
    if (this.pendingOrders.has(orderId)) {
      this.pendingOrders.delete(orderId);
      console.log(`   [PAPER] Cancelled Pending Order ${orderId}`);
    } else {
      console.log(
        `   [PAPER] Cancelled Order ${orderId} (Already filled or non-existent)`
      );
    }
  }

  async getOrderStatus(
    orderId: string
  ): Promise<{ matched: number; cancelled: boolean; avgFillPrice: number }> {
    // Check if it was filled
    if (this.filledOrders.has(orderId)) {
      const filled = this.filledOrders.get(orderId)!;
      return {
        matched: filled.matched,
        cancelled: false,
        avgFillPrice: filled.avgFillPrice,
      };
    }
    // Check if it's still in pending
    if (this.pendingOrders.has(orderId)) {
      return { matched: 0, cancelled: false, avgFillPrice: 0 };
    }
    // Not found - assume cancelled
    return { matched: 0, cancelled: true, avgFillPrice: 0 };
  }

  async getPositions(
    tokenIdYes: string,
    tokenIdNo: string
  ): Promise<{ yes: number; no: number }> {
    const posYes = this.positions.get(tokenIdYes);
    const posNo = this.positions.get(tokenIdNo);

    return {
      yes: posYes ? posYes.quantity : 0,
      no: posNo ? posNo.quantity : 0,
    };
    }

  // 🟢 CRITICAL FIX: Implement getOpenOrders for order recovery
  async getOpenOrders(marketSlug: string): Promise<Array<{
    id: string;
    side: "YES" | "NO";
    price: string;
    size: number;
  }>> {
    const orders: Array<{
      id: string;
      side: "YES" | "NO";
      price: string;
      size: number;
    }> = [];

    // Return all pending orders (they're "open" in paper trading)
    for (const [orderId, params] of this.pendingOrders.entries()) {
      // 🟢 Get side from token mapping, or infer from stored data
      // Since we don't have market context here, we'll use a heuristic:
      // In SniperLadder, all orders are BUY, and side is determined by tokenId
      // For now, return a placeholder - the bot will need to match by tokenId
      const side = this.tokenSideMap.get(params.tokenId) || "YES"; // Default to YES
      
      orders.push({
        id: orderId,
        side: side,
        price: params.price.toFixed(2),
        size: params.size,
      });
    }

    return orders;
  }

  /**
   * Manual Fill Trigger for Strategy
   */
  async executePaperFill(
    orderId: string,
    price: number,
    quantity: number
  ): Promise<void> {
    const params = this.pendingOrders.get(orderId);
    if (!params) {
      console.warn(`❌ [PAPER] Cannot fill unknown/cancelled order ${orderId}`);
      return;
    }

    // 1. SIMULATE LATENCY
    await new Promise((resolve) => setTimeout(resolve, 50));

    const cost = price * quantity;

    // 2. VALIDATE FUNDS
    if (params.side === "BUY" && this.balance < cost) {
      console.warn(`❌ [PAPER] REJECTED: Insufficient Funds`);
      this.pendingOrders.delete(orderId);
      return;
    }

    // 3. EXECUTE TRADE
    if (params.side === "BUY") {
      this.balance -= cost;
    } else {
      this.balance += cost; // Sell logic
    }

    // 4. UPDATE POSITIONS
    this.updatePosition({ ...params, price, size: quantity }); // Use fill price

    // 5. MARK AS FILLED (store actual fill price!)
    this.filledOrders.set(orderId, { matched: quantity, avgFillPrice: price });

    // 6. REMOVE FROM PENDING
    this.pendingOrders.delete(orderId);

    // 7. LOG
    this.printTradeLog(orderId, { ...params, price }, cost);
    this.printPortfolioSummary();
  }

  // --- INTERNAL TRACKING LOGIC ---

  private updatePosition(params: TradeParams) {
    const existing = this.positions.get(params.tokenId);

    if (existing) {
      // Calculate Weighted Average Price
      // New Avg = ((OldQty * OldAvg) + (NewQty * NewPrice)) / TotalQty
      const totalShares = existing.quantity + params.size;
      const totalCost = existing.totalInvested + params.price * params.size;

      existing.quantity = totalShares;
      existing.totalInvested = totalCost;
      existing.avgEntryPrice = totalCost / totalShares;
    } else {
      // New Position
      this.positions.set(params.tokenId, {
        tokenId: params.tokenId,
        side: params.side,
        quantity: params.size,
        avgEntryPrice: params.price,
        totalInvested: params.price * params.size,
      });
    }
  }

  private printTradeLog(id: string, params: TradeParams, cost: number) {
    console.log(`\n✅ [PAPER TRADE EXECUTED] #${id}`);
    console.log(`   🔹 Side:      ${params.side}`);
    console.log(`   🔹 Asset:     ...${params.tokenId.slice(-10)}`); // Show last 10 chars
    console.log(`   🔹 Price:     $${params.price.toFixed(2)}`);
    console.log(`   🔹 Size:      ${params.size} Shares`);
    console.log(`   🔹 Cost:      $${cost.toFixed(2)}`);
  }

  private printPortfolioSummary() {
    const totalInvested = Array.from(this.positions.values()).reduce(
      (sum, p) => sum + p.totalInvested,
      0
    );
    const totalEquity = this.balance + totalInvested;
    const profit = totalEquity - this.initialBalance;
    const profitColor = profit >= 0 ? "🟢" : "🔴";

    console.log(`\n📊 [PORTFOLIO SUMMARY]`);
    console.log(`   💵 Cash Balance:    $${this.balance.toFixed(2)}`);
    console.log(`   💼 Invested Amount: $${totalInvested.toFixed(2)}`);
    console.log(`   📈 Total Equity:    $${totalEquity.toFixed(2)}`);
    console.log(`   ${profitColor} PnL (Unrealized): $${profit.toFixed(2)}`);

    console.log(
      `   -----------------------------------------------------------`
    );
    console.log(`   POSITIONS:`);
    if (this.positions.size === 0) {
      console.log(`   (Empty)`);
    } else {
      this.positions.forEach((pos, id) => {
        console.log(
          `   • ID ...${id.slice(-8)} | Qty: ${
            pos.quantity
          } | Avg Entry: $${pos.avgEntryPrice.toFixed(
            3
          )} | Cost: $${pos.totalInvested.toFixed(2)}`
        );
      }      );
    }
    console.log(
      `   -----------------------------------------------------------\n`
    );
  }

  // 🟢 NEW: Get user address (stub for paper trading)
  async getAddress(): Promise<string> {
    return "0x0000000000000000000000000000000000000000"; // Dummy address for paper trading
  }

  // 🟢 NEW: Redeem winnings (stub for paper trading)
  async redeemPositions(conditionId: string): Promise<string> {
    console.log(`📝 [PAPER] Redemption simulated for condition ${conditionId.slice(0, 10)}...`);
    // In paper trading, redemption doesn't make sense - just return a dummy hash
    return `PAPER-REDEEM-${Date.now()}`;
  }
}
